### ISSUE TITLE
Potential use-after-free in SessionOptions

### ISSUE THREAD

**Issue by IvanUkhov** [google/tensorflow-rust#16]
Currently, `SessionOptions::set_target` looks as [follows](https://github.com/google/tensorflow-rust/blob/916e83a5dd6b549edffedae703505c2c2d8be8da/src/lib.rs#L278):

```rust
pub fn set_target(&mut self, target: &str) -> std::result::Result<(), NulError> {
  let cstr = try!(CString::new(target));
  unsafe {
    tf::TF_SetTarget(self.inner, cstr.as_ptr());
  }
  Ok(())
}
```

So `cstr` is dropped when `set_target` returns. However, `TF_SetTarget` doesn’t copy the string; it just [stores the pointer](https://github.com/tensorflow/tensorflow/blob/r0.9/tensorflow/core/client/tensor_c_api.cc#L177) in an internal structure:

```c
void TF_SetTarget(TF_SessionOptions* options, const char* target) {
  options->options.target = target;
}
```

Regards,
Ivan

**Comment by IvanUkhov** [comment]
I’ve found an answer to this question. The [type](https://github.com/tensorflow/tensorflow/blob/r0.9/tensorflow/core/public/session_options.h#L55) of `target` is a C++ `string`, which copies the memory via `operator= (const char *)`. Sorry for the false positive.
